ssm_parameter: add support for tags (#1573)

mikehas opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Adding support for tags following community guidelines and other practices from other modules. secretsmanager_secret was used along with helper functions from ec2 code. Addresses open issue for feature request #1573.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ssm_parameter

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

mikehas commented on plugins/modules/ssm_parameter.py:
```diff
@@ -208,12 +232,35 @@
+      description: A dictionary representing the tags associated with the parameter.
+      type: dict
+      returned: when the parameter has tags
+      example: {'MyTagName': 'Some Value'}
```
Hi Alina, I added version_added as requested.

mikehas commented on plugins/modules/ssm_parameter.py:
```diff
@@ -90,6 +90,7 @@
   - amazon.aws.ec2
   - amazon.aws.boto3
+  - amazon.aws.tags
 '''
```
@alinabuzachis, Would you like me to add this note about when tags support was added or should I wait for @markuman to chime in?

alinabuzachis replied:
I would add that note because that way we'll know exactly when tag support was added.

mikehas replied:
Thanks Alina, I updated the docs as suggested.
```
notes:
  - Support for I(tags) and I(purge_tags) was added in release 5.1.0.
```

dennisqian1 commented on plugins/modules/ssm_parameter.py:
```diff
@@ -387,7 +493,27 @@ def create_update_parameter(client, module):
+            # time.sleep(300)
+
+        # Handle tag updates for existing parameters
+        if (module.params.get('overwrite_value') != 'never'):
```
```suggestion
        if module.params.get('overwrite_value') != 'never':
```

mikehas replied:
Thanks Dennis, Done. I also removed the useless commented out code on line 497.

mikehas commented:
Hi @alinabuzachis, I think I've addressed all the comments so far. Is there anything else I can do to help increase the chances of this PR getting merged?

mikehas commented:
> Hi @alinabuzachis, I think I've addressed all the comments so far. Is there anything else I can do to help increase the chances of this PR getting merged?

cc @116davinder @jillr @markuman @mikedlr @nathanwebsterdotme @ozbillwang @s-hertel @tremble

116davinder commented:
LGTM

markuman commented on plugins/modules/ssm_parameter.py:
```diff
@@ -90,6 +90,11 @@
+  - amazon.aws.tags
+
+notes:
+  - Support for I(tags) and I(purge_tags) was added in release 5.1.0.
```
```suggestion
  - Support for I(tags) and I(purge_tags) was added in release 5.2.0.
```

markuman commented on plugins/modules/ssm_parameter.py:
```diff
@@ -208,12 +236,37 @@
+          type: str
+          example: Some value.
+          returned: success
+      version_added: 5.1.0
```
```suggestion
      version_added: 5.2.0
```

markuman commented on plugins/modules/ssm_parameter.py:
```diff
@@ -444,6 +572,8 @@ def setup_module_object():
         overwrite_value=dict(default='changed', choices=['never', 'changed', 'always']),
         tier=dict(default='Standard', choices=['Standard', 'Advanced', 'Intelligent-Tiering']),
+        tags=dict(type='dict', aliases=['resource_tags']),
+        purge_tags=dict(type='bool', default=True),
```
Imo this is a breaking change. Because until `community.aws 5.1.0` you can use `ssm_parameter` and it won't touch the tags. But now it would just purge your tags.  
This must be set to `default=False` for the moment of release.  

cc @alinabuzachis

mikehas replied:
Even with purge_tags=True as the default, existing tags will only be removed if the tags param is populated. To remove all tags, the empty dict needs to be supplied. I don't think this is a breaking change because someone would have had to set the tags parameter in existing tasks, which I believe would have thrown an error prior to this change.

markuman replied:
Ah right. thanks!

alinabuzachis commented on plugins/modules/ssm_parameter.py:
```diff
@@ -137,6 +142,29 @@
     msg: "{{ lookup('amazon.aws.aws_ssm', 'Hello') }}"
+
+- name: Create or update key/value pair in AWS SSM parameter store w/ tags
+  community.aws.ssm_paramater:
```
```suggestion
  community.aws.ssm_parameter:
```

markuman replied:
I think the PR must just be rebased now: https://github.com/ansible-collections/community.aws/pull/1644

alinabuzachis commented on plugins/modules/ssm_parameter.py:
```diff
@@ -137,6 +142,29 @@
+      Tag With Space: "foo bar"
+
+- name: Add or update a tag on an existing parameter w/o removing existing tags
+  community.aws.ssm_paramater:
```
```suggestion
  community.aws.ssm_parameter:
```

alinabuzachis commented on plugins/modules/ssm_parameter.py:
```diff
@@ -137,6 +142,29 @@
+      Contact: "person1"
+
+- name: Delete all tags on an existing parameter
+  community.aws.ssm_paramater:
```
```suggestion
  community.aws.ssm_parameter:
```

alinabuzachis commented on plugins/modules/ssm_parameter.py:
```diff
@@ -208,12 +236,37 @@
+      type: dict
+      returned: when the parameter has tags
+      example: {'MyTagName': 'Some Value'}
+      version_added: 5.1.0
```
```suggestion
      version_added: 5.2.0
```

alinabuzachis commented on plugins/modules/ssm_parameter.py:
```diff
@@ -208,12 +236,37 @@
       description: Parameter version number
       example: 3
       returned: success
+    tags:
```
@markuman What if we only return tags as a dictionary rather than a list of dicts and remove tags_dict from the response?

mikehas replied:
@markuman @alinabuzachis To provide a little context, tags_dict was copied over from plugins/modules/secretsmanager_secret.py. I'd be happy to remove the code if you'd like.

tremble replied:
secretsmanager_secret has some compatibility code in there because it originally returned the "boto3 style" (list of dicts) format, rather than the normal simple dictionary.

We often just convert the resource objects the APIs return from CamelCase to snake_case, and add that as part of what the module returns.  When AWS suddenly adds support for Tags to a resource and changes what the API returns, we start returning the list-of-dict style tags, and need to go through a deprecation cycle before we can return the simple dict.

Since returning tags is new to this module, we can skip the "list-of-dict" and just return the "dict" format as "tags".

tremble replied:
(There are also some cases where reviews simply missed that the module originally returned the boto3 style tags)
- We're trying to tidy this up :)

tremble commented on plugins/modules/ssm_parameter.py:
```diff
@@ -90,6 +90,11 @@
+  - amazon.aws.tags
+
+notes:
+  - Support for I(tags) and I(purge_tags) was added in release 5.2.0.
```
```suggestion
  - Support for I(tags) and I(purge_tags) was added in release 5.3.0.
```

tremble reviewed (changes requested):
Many thanks for this, and I'm sorry we missed responding in time to get this into 5.2.0.  For the sake of getting this merged I'll update the version information and drop the boto3-style tags.

tremble commented:
@mikehas thanks for taking the time to submit this PR, I'm sorry it got stuck for a while.
